pyexactreal/src/pyexactreal/cppyy_exactreal.py

```diff
@@ -29,76 +75,152 @@
+
+            implementation = getattr(cppyy.gbl.exactreal.cppyy, op)
+            def binary(lhs, rhs, implementation=implementation):
+                return unwrap_binary_optional(implementation[type(lhs), type(rhs)](lhs, rhs))
+            def rbinary(rhs, lhs, implementation=implementation):
+                return unwrap_binary_optional(implementation[type(lhs), type(rhs)](lhs, rhs))
+
+            setattr(proxy, python_op, binary)
+            setattr(proxy, python_rop, rbinary)
+
+        setattr(proxy, "__neg__", lambda self: cppyy.gbl.exactreal.cppyy.neg(self))
 
-def enable_arithmetic(proxy, name):
-    elements = ["Element<exactreal::IntegerRing>", "Element<exactreal::RationalField>", "Element<exactreal::NumberField>"]
-    with_precision = ["Arb", "Arf"]
-    if name in with_precision:
-        for (n, op) in [('add', ord('+')), ('sub', ord('-')), ('mul', ord('*')), ('div', ord('/'))]:
-            def create_expression(lhs, rhs, op = op): return Expression(lhs, rhs, op)
-            def inplace(lhs, *args, **kwargs): raise NotImplementedError("inplace operators are not supported")
-            setattr(proxy, "__%s__"%n, create_expression)
-            setattr(proxy, "__i%s__"%n, inplace)
-    if name in elements:
-        for (n, op) in [('add', ord('+')), ('sub', ord('-')), ('mul', ord('*')), ('div', ord('/'))]:
-            def cppname(x):
-                # some types such as int do not have a __cppname__; there might
-                # be a better way to get their cppname but this seems to work
-                # fine for the types we're using at least.
-                return type(x).__cppname__ if hasattr(type(x), '__cppname__') else type(x).__name__
-            def binary(lhs, rhs, op = op):
-                return cppyy.gbl.exactreal.boost_binary[cppname(lhs), cppname(rhs), op](lhs, rhs)
-            def inplace(lhs, *args, **kwargs): raise NotImplementedError("inplace operators are not supported yet")
-            setattr(proxy, "__%s__"%n, binary)
-            setattr(proxy, "__r%s__"%n, binary)
-            setattr(proxy, "__i%s__"%n, inplace)
-    if name in elements + with_precision:
-        # strangely, cppyy does not hook up the unary operator-, so we do that here manually
-        setattr(proxy, "__neg__", lambda self: cppyy.gbl.exactreal.minus(self))
-    if name == "Arf":
-        # cppyy has trouble with our nested enum in Arf (I could not reproduce
-        # it so it is unclear what is the reason for this) so we provide the
-        # enum values explicitly on the Arf instance.
-        proxy.NEAR = 4
-        proxy.DOWN = 0
-        proxy.UP = 1
-        proxy.FLOOR = 2
-        proxy.CEIL = 3
 
 cppyy.py.add_pythonization(enable_arithmetic, "exactreal")
+cppyy.py.add_pythonization(enable_arithmetic, "exactreal::yap")
+
+def enable_yap(proxy, name):
+    r"""
+    Make yap templating machinery accessible from cppyy.
+
+    TESTS::
+
+        >>> from pyexactreal import exactreal
```

Comment:
@videlec this also fixes Arb from Python I hope. Before I had some spurious problems with segfaults. Also the printing should now make it clear what's going on.